### ISSUE TITLE
Training code

### ISSUE THREAD

**Issue by sduscx** [YadiraF/PRNet#105]
Hi,as a newer to this, I failed to achieve the training code, could you please provide training code to me? I hope I can get your help, thank you very much. My email: m17864154809@163.com

**Comment by zhuixunforever** [comment]
Hi. I found a kind person doing this job. You can download it by search: https://github.com/jnulzl/PRNet, because of my bad computer, I have no chance to train the whole data, but I know it can run. So I don't know if his training code is right. If you tried it successfully, please tell me, thank you!
Best wishes to you.

**Comment by doujl** [comment]
Hi.I also need to train this network with a new training set. Don't know if you are successful?If you succeed, can you tell me how to do it?My email: liangdjsha@163.com,thank you!

**Comment by LucienXian** [comment]
@zhuixunforever Hi, I want to know the "--train_data_file“，how to generate it in the repo https://github.com/jnulzl/PRNet? Can you give some tips?

**Comment by zhuixunforever** [comment]
@LucienXian Hi, I generate train data by the author' face3d, the eighth example, [generate_posmap_300WLP](https://github.com/YadiraF/face3d/blob/master/examples/8_generate_posmap_300WLP.py), like the author said.

**Comment by WWWJL** [comment]
> Hi. I found a kind person doing this job. You can download it by search: https://github.com/jnulzl/PRNet, because of my bad computer, I have no chance to train the whole data, but I know it can run. So I don't know if his training code is right. If you tried it successfully, please tell me, thank you!
> Best wishes to you.

i have a question ,the train date label ，how to produce .npy?

**Comment by zhuixunforever** [comment]
@WWWJL I use the data that the author said, it's [300W_LP](http://www.cbsr.ia.ac.cn/users/xiangyuzhu/projects/3DDFA/main.htm), and use the author' code to produce .npy, it's [generate_posmap_300WLP](https://github.com/YadiraF/face3d/blob/master/examples/8_generate_posmap_300WLP.py)

**Comment by WWWJL** [comment]
ok,i got it ,thank you very much




------------------ 原始邮件 ------------------
发件人: "zhuixunforever"<notifications@github.com>;
发送时间: 2019年7月9日(星期二) 上午9:13
收件人: "YadiraF/PRNet"<PRNet@noreply.github.com>;
抄送: "吴佳玲"<1324244977@qq.com>;"Mention"<mention@noreply.github.com>;
主题: Re: [YadiraF/PRNet] Training code (#105)




@WWWJL I use the data that the author said, it's 300W_LP, and use the author' code to produce .npy, it's generate_posmap_300WLP

**Comment by zhuixunforever** [comment]
@WWWJL that's all right!

**Comment by WWWJL** [comment]
Hi，我一直有个疑惑，我们使用generate_posmap_300WLP脚本生成每幅图片的uv position 图，我看这个脚本的输入都是同样的BFM 模型(用matlab 运行generate.m保存在Out目录下的)。我认为每张图应该有它们各自的BFM 模型，再生成相应的un 位置图。这一部分您是如何理解的




------------------ 原始邮件 ------------------
发件人: "zhuixunforever"<notifications@github.com>;
发送时间: 2019年7月9日(星期二) 上午9:26
收件人: "YadiraF/PRNet"<PRNet@noreply.github.com>;
抄送: "吴佳玲"<1324244977@qq.com>;"Mention"<mention@noreply.github.com>;
主题: Re: [YadiraF/PRNet] Training code (#105)




@WWWJL that's all right!

**Comment by WWWJL** [comment]
噢。我懂了，这个是300W_LP每张图对应的mat信息，用它们来产生自己各自的BFM




------------------ 原始邮件 ------------------
发件人: "zhuixunforever"<notifications@github.com>;
发送时间: 2019年7月9日(星期二) 上午9:26
收件人: "YadiraF/PRNet"<PRNet@noreply.github.com>;
抄送: "吴佳玲"<1324244977@qq.com>;"Mention"<mention@noreply.github.com>;
主题: Re: [YadiraF/PRNet] Training code (#105)




@WWWJL that's all right!

**Comment by zhuixunforever** [comment]
@WWWJL 是的嘞，每个mat里面包含了这个图（这个人）的3维信息，位姿信息，你按照作者代码 的要求把它需要的文件放在指定 地方，然后再加载300W_LP的数据就可以得到npy了

**Comment by WWWJL** [comment]
嗯嗯，这样看，这篇paper的工作量还挺小的，端到端只生成uv map吗？视频里的面具贴合以及人脸关键点输出是否需要其他途径实现呢？(下图)




------------------ 原始邮件 ------------------
发件人: "zhuixunforever"<notifications@github.com>; 
发送时间: 2019年7月9日(星期二) 上午10:57
收件人: "YadiraF/PRNet"<PRNet@noreply.github.com>; 
抄送: "吴佳玲"<1324244977@qq.com>; "Mention"<mention@noreply.github.com>; 
主题: Re: [YadiraF/PRNet] Training code (#105)




@WWWJL 是的嘞，每个mat里面包含了这个图（这个人）的3维信息，位姿信息，你按照作者代码 的要求把它需要的文件放在指定 地方，然后再加载300W_LP的数据就可以得到npy了

**Comment by zhuixunforever** [comment]
那个uv_map已经是经过贴合的，然后人脸关键点其实也是在固定的位置。作者用那个纹理贴图的那个uv来作为3D到2D的相互转换，所以如果恢复出了uv_map，也就能够恢复出3D模型，并且这个3D模型还和图像人脸对应。

**Comment by zhuixunforever** [comment]
工作量比较小，但是我觉得很新颖，而且最后效果也不错

**Comment by WWWJL** [comment]
那使用uv_map来完成下面这些效果，相应代码有吗？



------------------ 原始邮件 ------------------
发件人: "zhuixunforever"<notifications@github.com>;
发送时间: 2019年7月9日(星期二) 中午11:12
收件人: "YadiraF/PRNet"<PRNet@noreply.github.com>;
抄送: "吴佳玲"<1324244977@qq.com>;"Mention"<mention@noreply.github.com>;
主题: Re: [YadiraF/PRNet] Training code (#105)




那个uv_map已经是经过贴合的，然后人脸关键点其实也是在固定的位置。作者用那个纹理贴图的那个uv来作为3D到2D的相互转换，所以如果恢复出了uv_map，也就能够恢复出3D模型，并且这个3D模型还和图像人脸对应。

**Comment by WWWJL** [comment]
对uv_map的使用我还不是很清楚，你们的意思是只要得到了uv_map，这些关键点和面具贴合都可以实现。现在，我的问题是，如何使用uv_map来实习后面这些功能




------------------ 原始邮件 ------------------
发件人: "zhuixunforever"<notifications@github.com>;
发送时间: 2019年7月9日(星期二) 中午11:12
收件人: "YadiraF/PRNet"<PRNet@noreply.github.com>;
抄送: "吴佳玲"<1324244977@qq.com>;"Mention"<mention@noreply.github.com>;
主题: Re: [YadiraF/PRNet] Training code (#105)




那个uv_map已经是经过贴合的，然后人脸关键点其实也是在固定的位置。作者用那个纹理贴图的那个uv来作为3D到2D的相互转换，所以如果恢复出了uv_map，也就能够恢复出3D模型，并且这个3D模型还和图像人脸对应。

**Comment by zhuixunforever** [comment]
生成label的代码你可以看我最开始给你发的。然后作者这篇文章的工作的代码，只有模型，没有训练代码，我找过别人的代码， 还阔以

**Comment by WWWJL** [comment]
嗯嗯，我有训练代码，生成lable的方法也知道。现在的疑惑是，我训练好模型，用它去test,输出的是image 的uv图，怎么使用它来得到刚发的那种效果图。


------------------ 原始邮件 ------------------
发件人: "zhuixunforever"<notifications@github.com>;
发送时间: 2019年7月9日(星期二) 中午11:21
收件人: "YadiraF/PRNet"<PRNet@noreply.github.com>;
抄送: "吴佳玲"<1324244977@qq.com>;"Mention"<mention@noreply.github.com>;
主题: Re: [YadiraF/PRNet] Training code (#105)




生成label的代码你可以看我最开始给你发的。然后作者这篇文章的工作的代码，只有模型，没有训练代码，我找过别人的代码， 还阔以

**Comment by zhuixunforever** [comment]
你说的后面的功能是指什么？？？作者在README.md提到的那些应用嘛，比如Face Alignment，3D Face Reconstruction，3D Pose Estimation，Depth image，Texture Editing？？？

**Comment by zhuixunforever** [comment]
我其实没有看到你发的效果图。然后作者的代码里面有写3d点到obj里面，那个函数之前的函数就是把uv_map转化成3D点

**Comment by WWWJL** [comment]
https://github.com/YadiraF/PRNet
就是这里面的效果图，嗯嗯。我有用作者提供的模型测试，生成了obj,用meshlab打开是3D人脸。但是像链接里发的那个视频里的效果是如何实现的？




------------------ 原始邮件 ------------------
发件人: "zhuixunforever"<notifications@github.com>;
发送时间: 2019年7月9日(星期二) 中午11:36
收件人: "YadiraF/PRNet"<PRNet@noreply.github.com>;
抄送: "吴佳玲"<1324244977@qq.com>;"Mention"<mention@noreply.github.com>;
主题: Re: [YadiraF/PRNet] Training code (#105)




我其实没有看到你发的效果图。然后作者的代码里面有写3d点到obj里面，那个函数之前的函数就是把uv_map转化成3D点

**Comment by zhuixunforever** [comment]
那个视频不就是把68个关键点，3维人脸模型，还有一个不知道算啥，以视频的方式显示出来嘛？然后作者这个方法本身也很快，在gpu下做到实时也阔以

**Comment by WWWJL** [comment]
嗯嗯，模型输出有_mesh.mat,.txt,.obj文件。我看看把这些信息画在图片上，应该就是那个效果




------------------ 原始邮件 ------------------
发件人: "zhuixunforever"<notifications@github.com>;
发送时间: 2019年7月9日(星期二) 中午11:47
收件人: "YadiraF/PRNet"<PRNet@noreply.github.com>;
抄送: "吴佳玲"<1324244977@qq.com>;"Mention"<mention@noreply.github.com>;
主题: Re: [YadiraF/PRNet] Training code (#105)




那个视频不就是把68个关键点，3维人脸模型，还有一个不知道算啥，以视频的方式显示出来嘛？然后作者这个方法本身也很快，在gpu下做到实时也阔以

**Comment by zhuixunforever** [comment]
差不多，对的

**Comment by WWWJL** [comment]
嗨，我还是不大清楚，效果图里是怎么把人脸模型贴到2D的人脸图像上的？我现在点云数据，它还需要经过3d->2d投影才能贴到2d人脸上吗？




------------------ 原始邮件 ------------------
发件人: "zhuixunforever"<notifications@github.com>;
发送时间: 2019年7月9日(星期二) 中午11:34
收件人: "YadiraF/PRNet"<PRNet@noreply.github.com>;
抄送: "吴佳玲"<1324244977@qq.com>;"Mention"<mention@noreply.github.com>;
主题: Re: [YadiraF/PRNet] Training code (#105)




你说的后面的功能是指什么？？？作者在README.md提到的那些应用嘛，比如Face Alignment，3D Face Reconstruction，3D Pose Estimation，Depth image，Texture Editing？？？

**Comment by zhuixunforever** [comment]
加微信吧，zhuiXunForeverXly

**Comment by qiannvMao** [comment]
@WWWJL   @zhuixunforever hi. 我生成了.npy文件，怎样把它处理成训练数据 "--train_data_file“，您能把生成 "--train_data_file“的代码和训练代码发我一份不。邮箱：1452158355@qq.com

**Comment by qiannvMao** [comment]
@zhuixunforever @WWWJL 您好，能否将您的训练代码发我一份，我刚学习深度学习，复现训练代码出现问题，非常感谢。邮箱：mqn20170910@163.com

**Comment by yexiaoxi01** [comment]
> https://github.com/YadiraF/PRNet 就是这里面的效果图，嗯嗯。我有用作者提供的模型测试，生成了obj,用meshlab打开是3D人脸。但是像链接里发的那个视频里的效果是如何实现的？
> […](#)
> ------------------ 原始邮件 ------------------ 发件人: "zhuixunforever"<notifications@github.com>; 发送时间: 2019年7月9日(星期二) 中午11:36 收件人: "YadiraF/PRNet"<PRNet@noreply.github.com>; 抄送: "吴佳玲"<1324244977@qq.com>;"Mention"<mention@noreply.github.com>; 主题: Re: [YadiraF/PRNet] Training code (#105) 我其实没有看到你发的效果图。然后作者的代码里面有写3d点到obj里面，那个函数之前的函数就是把uv_map转化成3D点 — You are receiving this because you were mentioned. Reply to this email directly, view it on GitHub, or mute the thread.


你好，我想问问你有没有用自己训练的模型生成.obj. 我用meshlab打开的face shape 和 纹理贴图不匹配。不知道为什么。还有你对数据集里面的.mat文件了解吗， 他如何和BFM模型生成uv_posmap的,和3DMM模型是什么关系。虽然可以通过作者提供的代码生成uv图，但是原理很模糊。

**Comment by zhuixunforever** [comment]
@yexiaoxi01 噢，我忘说了，我当时跑那个代码，效果不好，最后哦纹理对不上，后来又找了一个，https://github.com/shen1994/DeepFaceRestruction，emmm，这个我跑的效果还阔以

**Comment by yexiaoxi01** [comment]
@zhuixunforever  非常感谢你的回复，我会试着用这个来训练模型。还有，你对mat文件里面的信息和bfm模型如何应用了解吗。我一直对生成uv图的这个过程比较困惑。这个3dMM模型和bfm模型是怎么用的。再次感谢你的帮助*-*

**Comment by zhuixunforever** [comment]
@yexiaoxi01 它那个mat里面存了很多关于当前人脸的信息的参数，包括3维，颜色，关键点等等，对于这篇文章来说，你简单理解为，通过.mat信息再加上其他的公共部分信息可以得到这个.mat代表的人脸的3D点就好了，然后这个3D点和图片是可以对应上的，作者就是把3D点放缩到图像大小（256），然后作者把3D点按照纹理贴图展开的方式展到二维上（普通3维的纹理展开到二维，你可以想象一下地球仪和2D平面地图，最后二维上某个像素保存的是其对应的3D点的rgb值），二维上某个像素保存的是其对应的3D点的xyz值（这个我觉得是作者这篇文章的idea），展开后的二维图就是那个uv图

**Comment by yexiaoxi01** [comment]
@zhuixunforever  谢谢解惑啦，我先尝试进行训练，有问题再问你。谢谢谢谢

**Comment by yexiaoxi01** [comment]
@zhuixunforever 你好，我是yexiaoxi01，在忙吗？之前请教你的那个问题，你推荐https://github.com/shen1994/DeepFaceRestruction 这个代码训练，我发现训练所有图像，很慢，大概需要近二十多天，还有就是，我如何通过训练好的模来设生成。obj文件，就像原作者那个run_basic.py. 我尝试好多修改，但是模型加载不出来，一直有问题。

**Comment by zhuixunforever** [comment]
@yexiaoxi01 加微信吧，zhuiXunForeverXly

**Comment by tiancaipipi110** [comment]
@zhuixunforever 想了一天到底作者如何从2d图转成3d .obj, 原来.mat里面有3d info... 这个.mat里面应该有vertice xyz, kpts, triangles, uv map, etc. 

> 生成label的代码你可以看我最开始给你发的

你说的label就是kpt的label吧，这个代码你还有吗？

> 展开后的二维图就是那个uv图

predictor就是用<a href="https://github.com/tiancaipipi110/PRNet/blob/master/predictor.py#L70">conv2d_transpose</a>的tensor回溯到3d的？

**Comment by vitahsu** [comment]
您好! 請問paper內部提到的 "we use the parameterized UV coordinates which computes a Tutte embedding with conformal Laplacian weight and then maps the mesh boundary to a square" 有沒有詳細的code描述BFM_UV.mat是怎麼生成的?
謝謝!

**Comment by zhao181** [comment]
> 您好! 請問paper內部提到的 "we use the parameterized UV coordinates which computes a Tutte embedding with conformal Laplacian weight and then maps the mesh boundary to a square" 有沒有詳細的code描述BFM_UV.mat是怎麼生成的?
> 謝謝!



您好同学！ 请问你有没有找到详细资料介绍BFM_UV.mat生成的?

**Comment by vitahsu** [comment]
@zhao181 在3D morphable models as spatial transformer networks這篇paper的section2.3 output grid是此篇paper引用的原出處，可以看一下! 不過看過去也還是非常模糊啊!!!!! (可能背景知識還不足) 若有問題歡迎一起討論! [另外在此連結也有一些討論唷!](https://github.com/anilbas/3DMMasSTN/issues/10#issuecomment-554349454) 謝謝

**Comment by zhao181** [comment]
谢谢 好的
On 11/21/2019 11:50，vitahsu<notifications@github.com> wrote：

@zhao181 在3D morphable models as spatial transformer networks這篇paper的section2.3 output grid是此篇paper引用的原出處，可以看一下! 不過看過去也還是非常模糊啊!!!!! (可能背景知識還不足) 若有問題歡迎一起討論! 另外在此連結也有一些討論唷! 謝謝

**Comment by vitahsu** [comment]
@WWWJL 
您好! 我最近使用tensorflow在復現paper，但是復現的磕磕拌拌的，想請問您是否願意分享您的訓練碼? 我的信箱為bantiauo@gmail.com，謝謝您!
另外若不方便分享的話想請教您，訓練300WLP這包6W多張Image在GPU上training的時間大概多久?謝謝!

**Comment by cuixin1992** [comment]
@vitahsu @zhao181 您好！请问您们现在明白如何生成BFM_UV.mat了吗？
